Implement CLI commands: planet data asset-get, download, activate, wait

kevinlacaille opened the pull request:
**Related Issue(s):**

Closes #505, #506, #507


**Proposed Changes:**

For inclusion in changelog (if applicable):

1. Add planet data asset-activate, asset-wait, asset-download CLI commands to help with activating and downloading assets.

Not intended for changelog:

1. 

**Diff of User Interface**

Old behavior:

N/A

New behavior:

**asset-activate**
```console
planet data asset-activate PSScene 20221003_002705_38_2461 basic_udm2
```

**asset-wait**
```console
planet data asset-wait PSScene 20221003_002705_38_2461 basic_udm2
00:00 - order my asset - state: active
{'_links': {'_self': 'https://api.planet.com/data/v1/assets/eyJpIjogIjIwMjIxMDAzXzAwMjcwNV8zOF8yNDYxIiwgImMiOiAiUFNTY2VuZSIsICJ0IjogImJhc2ljX3VkbTIiLCAiY3QiOiAiaXRlbS10eXBlIn0', 'activate': 'https://api.planet.com/data/v1/assets/eyJpIjogIjIwMjIxMDAzXzAwMjcwNV8zOF8yNDYxIiwgImMiOiAiUFNTY2VuZSIsICJ0IjogImJhc2ljX3VkbTIiLCAiY3QiOiAiaXRlbS10eXBlIn0/activate', 'type': 'https://api.planet.com/data/v1/asset-types/basic_udm2'}, '_permissions': ['download'], 'expires_at': '2023-03-01T23:25:19.908175', 'location': 'https://api.planet.com/data/v1/download?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzUxMiJ9.eyJzdWIiOiI0c0NHUW1rcXk4REg1RUpnYS1mV2RGRHp6VlJiT2VtdzFEVjU1OFpPQzE3ZE41THJlcmtxVzJ4c2J0T0t3bEREaWttenAzeHVPZ0w3OU1YNkQ1ZUxaZz09IiwiZXhwIjoxNjc3NzEzMTE5LCJ0b2tlbl90eXBlIjoidHlwZWQtaXRlbSIsIml0ZW1fdHlwZV9pZCI6IlBTU2NlbmUiLCJpdGVtX2lkIjoiMjAyMjEwMDNfMDAyNzA1XzM4XzI0NjEiLCJhc3NldF90eXBlIjoiYmFzaWNfdWRtMiJ9.i4ynH7e8xz-Sv6bUtYETKkujEczooMWD3p65Qf3am2OHTv6amnirBjSySsSTfG1gPcHz0smpNK_QvXD6108COA', 'md5_digest': '3a9f7dd1ce500f699d0a96afdd0e3aa2', 'status': 'active', 'type': 'basic_udm2'}
```

**asset-download**
```console
planet data asset-download PSScene 20221003_002705_38_2461 basic_udm2 --overwrite --checksum                 
/Users/kevin.lacaille/code/repos/planet/planet-client-python/20221003_002705_38_2461_1A_udm2.tif: 100%|████████████████████████████| 3.16k/3.16k [00:00<00:00, 26.5MB/s]
```

**Putting it all together**
```console
planet data asset-activate PSScene 20221003_002705_38_2461 basic_udm2 && \
planet data asset-wait PSScene 20221003_002705_38_2461 basic_udm2 && \
planet data asset-download PSScene 20221003_002705_38_2461 basic_udm2 --overwrite
00:00 - order my asset - state: active
{'_links': {'_self': 'https://api.planet.com/data/v1/assets/eyJpIjogIjIwMjIxMDAzXzAwMjcwNV8zOF8yNDYxIiwgImMiOiAiUFNTY2VuZSIsICJ0IjogImJhc2ljX3VkbTIiLCAiY3QiOiAiaXRlbS10eXBlIn0', 'activate': 'https://api.planet.com/data/v1/assets/eyJpIjogIjIwMjIxMDAzXzAwMjcwNV8zOF8yNDYxIiwgImMiOiAiUFNTY2VuZSIsICJ0IjogImJhc2ljX3VkbTIiLCAiY3QiOiAiaXRlbS10eXBlIn0/activate', 'type': 'https://api.planet.com/data/v1/asset-types/basic_udm2'}, '_permissions': ['download'], 'expires_at': '2023-03-02T19:30:48.942718', 'location': 'https://api.planet.com/data/v1/download?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzUxMiJ9.eyJzdWIiOiJQYWtGNHZuUEs3WXRmSFNGUklHY2I3YTNXT3piaTlaam4zWUpZMmxnd0x5cVlFMVBRSHU5QXNCcjR5Q3FxSjBNbl9yN3VwVEFQYUI1ZzhYNUJmcDhmUT09IiwiZXhwIjoxNjc3Nzg1NDQ4LCJ0b2tlbl90eXBlIjoidHlwZWQtaXRlbSIsIml0ZW1fdHlwZV9pZCI6IlBTU2NlbmUiLCJpdGVtX2lkIjoiMjAyMjEwMDNfMDAyNzA1XzM4XzI0NjEiLCJhc3NldF90eXBlIjoiYmFzaWNfdWRtMiJ9.Dd0opDjW3bBS6qLLZoNiJkfBsO2n5Xz9pM5apEUz_K6viDPFexhJiy6bMbaySbby8W0YvuATdb1uYXS2FkweDg', 'md5_digest': '3a9f7dd1ce500f699d0a96afdd0e3aa2', 'status': 'active', 'type': 'basic_udm2'}
/Users/kevin.lacaille/code/repos/planet/planet-client-python/20221003_002705_38_2461_1A_udm2.tif: 100%|██████████████████████████████████| 3.16k/3.16k [00:00<00:00, 32.0MB/s]
```

**PR Checklist:**

- [x] This PR is as small and focused as possible
- [x] If this PR includes proposed changes for inclusion in the changelog, the title of this PR summarizes those changes and is ready for inclusion in the Changelog.
- [x] I have updated docstrings for function changes and docs in the 'docs' folder for user interface / behavior changes
- [x] This PR does not break any examples or I have updated them

**(Optional) @mentions for Notifications:**

jreiberkyle commented on planet/cli/data.py:
```diff
@@ -399,8 +402,109 @@ async def search_update(ctx,
+@click.argument("asset", type=types.JSON())
+@click.option('--directory',
+              default='.',
+              help=('Base directory for file download.'),
```
fyi the parentheses aren't needed unless the string spans multiple lines

jreiberkyle commented on planet/cli/data.py:
```diff
@@ -399,8 +402,109 @@ async def search_update(ctx,
+@click.argument("item_id")
+@click.argument("asset_type_id")
+@pretty
+async def asset_get(ctx, item_types, item_id, asset_type_id, pretty):
```
now I understand why you wrote the get_asset cli command! I was thinking activate/wait asset would start with item_types/id/asset_id but yeah, this breaking it up and using piping makes sense. I'm hoping there's some documentation on usage added to the CLI docs 👀

jreiberkyle replied:
I think an asset is associated with only one item_type, no? If that's true then item_types -> item_type would make sense here.

jreiberkyle commented on planet/cli/data.py:
```diff
@@ -399,8 +402,109 @@ async def search_update(ctx,
+    '''Get an item asset.'''
+    async with data_client(ctx) as cl:
+        asset = await cl.get_asset(item_types, item_id, asset_type_id)
+        echo_json(asset, pretty)
```
ok this is a nit since it absolutely works either way, but echo_json doesn't need to be within the context of data_client and we have been calling it outside the context (at least when I look at the orders cli)
```suggestion
    echo_json(asset, pretty)
```

jreiberkyle commented on tests/integration/test_data_cli.py:
```diff
@@ -779,12 +785,192 @@ def test_search_update_fail(invoke, search_id, search_filter):
+        asset_download_command = [
+            'asset-download',
+            json.dumps(basic_udm2_asset),
+            '--directory',
+            Path(folder),
```
these could be a little easier to read with options specified like this:

```suggestion
            f'--directory={Path(folder)}',
```
not necessary, just an fyi

kevinlacaille replied:
This is way better, thanks!

jreiberkyle commented on tests/integration/test_data_cli.py:
```diff
@@ -779,12 +785,192 @@ def test_search_update_fail(invoke, search_id, search_filter):
+    runner = CliRunner()
+    with runner.isolated_filesystem() as folder:
+        if exists:
+            Path(folder, 'img.tif').write_text('i exist')
```
huh kind of interesting writing text into a tif

jreiberkyle commented on tests/integration/test_data_cli.py:
```diff
@@ -779,12 +785,192 @@ def test_search_update_fail(invoke, search_id, search_filter):
+    }
+
+    page_response = {
+        "basic_analytic_4b": {
+            "_links": {
+                "_self":
+                "SELFURL",
+                "activate":
+                "ACTIVATEURL",
+                "type":
+                "https://api.planet.com/data/v1/asset-types/basic_analytic_4b"
+            },
+            "_permissions": ["download"],
+            "md5_digest": None,
+            "status": "inactive",
+            "type": "basic_analytic_4b"
+        },
```
I'm curious why this part was included, I don't see where it's being used by the test.

kevinlacaille replied:
`page_response` is being used by `mock_resp` on line 958. i.e., 
`mock_resp = httpx.Response(HTTPStatus.OK, json=page_response)`

jreiberkyle reviewed (changes requested):
EDIT: 
Ok, wait, as I am looking at the usage in the PR desc I'm seeing an issue. If you save asset.json, then use `planet data asset-download asset.json`, you're not going to be successful because `asset.json` has not been updated since the asset was activated, and so it won't have the download urls. OK, let's go back to how the CLI-data docs specify the UI (that is, [design-docs/CLI-Data.md](https://github.com/planetlabs/planet-client-python/blob/main/design-docs/CLI-Data.md#asset-activate)), aka, input for the CLI commands are the item-type, item-id, and asset-type. This matches the orders UI as well.

INITIAL:
Ok so I tested this and found it to be a delight to use. The comments I have in the review are mostly fyis and style I think, I can't remember any necessary changes. 

The one thing that DOES need to change is `design-docs/CLI-Data.md` since the UI in this PR differs. Ideally, we'd be working from that doc and have discussed the different UI before you got to work here, but I'm not sure I called that out in the ticket for one, and for two, I am on board with this change. So, it should be a quick change but it's one we should get in there before moving forward. I'll approve when it's in!

jreiberkyle commented:
Ok, wait, as I am looking at the usage in the PR desc I'm seeing an issue. If you save asset.json, then use `planet data asset-download asset.json`, you're not going to be successful because `asset.json` has not been updated since the asset was activated, and so it won't have the download urls. OK, let's go back to how the CLI-data docs specify the UI, that is, input for the CLI commands are the item-type, item-id, and asset-type. This matches the orders UI as well.

kevinlacaille commented:
@jreiberkyle alternatively, perhaps a better way of doing it would be:
`planet data asset-get PSScene 20221003_002705_38_2461 basic_udm2 | planet data asset-download -`

kevinlacaille commented on planet/cli/data.py:
```diff
@@ -14,17 +14,20 @@
+from planet.reporting import StateBar
 from planet import data_filter, DataClient, exceptions
-from planet.clients.data import (LIST_SEARCH_TYPE,
+from planet.clients.data import (SEARCH_SORT,
```
These are mixed up due to some weird merge conflict resolution, sorry about that :)

kevinlacaille commented:
@jreiberkyle Here’s what I did since you last reviewed it:
- Removed `asset` as an input to `asset-activate`, `asset-wait`, `asset-download`, and replaced it with `item_id`, `item_type`, and `asset_type`
    - This got `get_asset()` involved in all three functions
- Changed the tests for all three CLI functions to accommodate the new parameters
    - Created `get_asset()` ’s mocked response in a pytest fixture, removed a lot of clutter
- Commented out `asset-get` from the CLI and tests
- Added `data asset` commands to the docs

jreiberkyle reviewed (changes requested):
Looks good, just the `item-type` type needs to change to str (one entry) instead of comma-separated string (multiple entries)

For example, this is confusing to me:
```
planet data asset-activate PSScene,SkySatScene 20221003_002705_38_2461 basic_udm2
Error: {"general": [{"message": "Could not find the requested item."}], "field": {}}
```

For 1 how am I ordering the same ID/asset for two different satellites and for 2 the error message isn't telling me anything about how I specified two item types but only one will be used, and which one was used, which would help me understand the source of this error (maybe... I can't say I'm in love with the API error either).

kevinlacaille commented on planet/cli/data.py:
```diff
@@ -480,8 +483,139 @@ async def search_update(ctx,
+@click.pass_context
+@translate_exceptions
+@coro
+@click.argument("item_type",
```
I agree about the string, but are you saying we don't want to validate `item_type` with the callback function? The response from the server for an unsupported `item_type` is not very helpful.

kevinlacaille replied:
However, handing off to the callback function with `item_type` as a string does cause issues, because it cycles through `item_type in item_types`, so handing off something like `"PSScene"` raises a `click.BadParameter()` because "P" isn't a valid `item_type`.

kevinlacaille replied:
Went with option 2, see: https://github.com/planetlabs/planet-client-python/pull/864/commits/f01a519a1addcbd2ddd8e2ee64ab1fa33a168647

jreiberkyle commented on planet/cli/data.py:
```diff
@@ -78,6 +81,18 @@ def check_item_types(ctx, param, item_types) -> Optional[List[dict]]:
+        raise click.BadParameter("Only provide 1 item type.")
+    try:
+        validate_item_type(item_type)
+        return item_type
+    except SpecificationException as e:
+        raise click.BadParameter(str(e))
+
+
```
Best practice is to move the return outside of the try/except block. This is for two reasons: 1. minimize the number of steps inside the `try` and 2. it's a little more readable when the return is at the top level and there is only one return.

```suggestion
    except SpecificationException as e:
        raise click.BadParameter(str(e))

    return item_type
```

jreiberkyle commented on planet/cli/data.py:
```diff
@@ -480,8 +495,133 @@ async def search_update(ctx,
+@coro
+@click.argument("item_type", type=str, callback=check_item_type)
+@click.argument("item_id")
+@click.argument("asset_type_id")
```
OK this is supa confusing but if you look at https://developers.planet.com/docs/apis/data/reference/#tag/Items-and-Assets/operation/ListItemAssets then the argument here (and everywhere else in these CLI commands) should be named `asset_type`

Sorry I missed this last time.

kevinlacaille replied:
I've been writing it as `asset_type_id` because that's what we use for those functions in the client, but `asset_type` makes way more sense, I'm on it.

jreiberkyle commented on planet/cli/data.py:
```diff
@@ -78,6 +81,18 @@ def check_item_types(ctx, param, item_types) -> Optional[List[dict]]:
+def check_item_type(ctx, param, item_type) -> Optional[List[dict]]:
+    '''Validates the item type provided by comparing it to all supported
+    item types.'''
+    if len(item_type.split(",")) > 1:
+        raise click.BadParameter("Only provide 1 item type.")
```
OK so I tried this:

```
> planet data asset-activate PSScene,SkySatScene 20221003_002705_38_2461 basic_udm2
Usage: planet data asset-activate [OPTIONS] ITEM_TYPE ITEM_ID ASSET_TYPE_ID
Try 'planet data asset-activate --help' for help.

Error: Invalid value for 'ITEM_TYPE': Only provide 1 item type.
```
And it's not totally clear to me that it is the comma in my value for item-type that caused the issue here. 

When I remove these lines of code and run it again I get:

```
> planet data asset-activate PSScene,SkySatScene 20221003_002705_38_2461 basic_udm2
Usage: planet data asset-activate [OPTIONS] ITEM_TYPE ITEM_ID ASSET_TYPE_ID
Try 'planet data asset-activate --help' for help.

Error: Invalid value for 'ITEM_TYPE': item_type - 'PSScene,SkySatScene' is not one of 'MYD09GQ', 'MYD09GA', 'MOD09GA', 'Landsat8L1G', 'REOrthoTile', 'Sentinel2L1C', 'SkySatCollect', 'SkySatScene', 'REScene', 'PSScene', 'MOD09GQ', 'PSOrthoTile', 'Sentinel1', 'PSScene4Band', 'PSScene3Band'.
```
and to be honest, the second error code gives me more of the information I need to fix the problem.

So my recc here would be to just delete this check.

```suggestion
```

jreiberkyle reviewed (changes requested):
so close, so close!  just three changes requested, two can be as simple as accepting suggestions, the other is an argument name change.

jreiberkyle reviewed (approved):
lgtm!
